### ISSUE TITLE
Use web-based protocol handlers to make following across instances easier

### ISSUE THREAD

**Issue by jmaris** [mastodon/mastodon#19679]
### Pitch

Use [web-based protocol handlers](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler/Web-based_protocol_handlers) to automatically direct follow requests to the users instance.

- When the User logs in to their instance for the first time, they are given an opportunity to register the mastodon:// protocol to open with their instance.
- When the user clicks a follow button on an instance outside of their own, the button links to mastodon://user-to-be-followed@external-instance.com/follow
- This link is opened in the users instance automatically, resolves to the profile and gives the user the opportunity to confirm they would like to follow the account

### Motivation

The experience of trying to follow someone from another instance is currently very confusing, especially for new users. This would significantly simplify the process while preserving security/privacy.

**Comment by Gargron** [comment]
We've done this before but removed because browser support / UX was inadequate.

**Comment by Gottox** [comment]
## status quo

The workflow of following users from different instances is weird.

1. Visit a profile on a different instance
2. click on a follow button
3. be presented by a UI that can be different depending on the instance you're on[1]
  Depending on the workflow of the server:
    * for edi.social:
      1. find the right textbox to put in your user name and instance without any way to autocomplete (it seems to be prefilled, if you followed s/o on that instance before)
      2. click follow
    * for mastodon:
      1. copy the link to the profile
      2. go to your instance
      3. insert the link into the search field
      4. wait for the search results
      5. click the follow button 

## protocol handlers

With protocol handlers this can be simplified to:

* once: install protocol handler get the weird UI *once*

* afterwards:
  1. click the follow button
  3. confirm follow on your instance

It seems that there's a way to check [if a protocol handler is installed](https://github.com/vireshshah/custom-protocol-check), although it seems to be hacky. So it could be possible to even fall back to the old workflow.

## Appendix

[1]:
* mastodon.social ![image](https://user-images.githubusercontent.com/1056976/200111599-a3b7ddf8-8b07-4a35-a1c8-5a7974ad5d99.png)
* edi.social
![image](https://user-images.githubusercontent.com/1056976/200111642-f3c29931-4730-413e-b08c-8e27a85d61c4.png)

**Comment by Gottox** [comment]
This is just a collection of ideas after playing with the [`Navigator.registerProtocolHandler()`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler)  API.

1. insert testing iframe with something like `web+mastodon:ping`
2. the iframe will emit a "load" event if it resolves to a homepage. If not, the old workflow can be used.
3. If the user clicks on the follow button, it can be replaced by a iframe pointing to `web+mastodon:follow/<USERNAME>@<INSTANCE>`
4. This will show an confirmation dialog or show an "already followed" label
5. if the clicks on the confirmation, the account follows.

The same mechanism can be used to implement Social buttons and include them in websites.

Just an idea that needed to be written down :D

**Comment by jmaris** [comment]
@Gottox I think for new users the process of entering your instance every time you want to follow someone is frustrating. Especially when there is a solution that allows you to avoid that issue, that is why using protocol handlers is still significantly better than the edi.social approach.

**Comment by veale** [comment]
It seems the previous removal was based on UX scaring people when the handler was not opt-in. Opt-in was dismissed on the basis of 'too many opt in features' in Mastodon (https://github.com/mastodon/mastodon/pull/8127#issuecomment-411038820), which I don't think is a particularly convincing reason given the huge time save this provides, and the push it would make to actually improve the standard and browsers' handling of it. This should be (re)introduced as an opt-in feature.

**Comment by arivero** [comment]
Just curious, was the protocol handler fully removed or only the registerProtocolHandler call? Because I see that there is still a intent controller that seems to check for 'web+mastodon' uri scheme

https://github.com/mastodon/mastodon/blob/main/app/controllers/intents_controller.rb
https://github.com/mastodon/mastodon/blob/main/spec/controllers/intents_controller_spec.rb

but I can not find any code actually registering the scheme

**Comment by jmaris** [comment]
> It seems the previous removal was based on UX scaring people when the handler was not opt-in. Opt-in was dismissed on the basis of 'too many opt in features' in Mastodon ([#8127 (comment)](https://github.com/mastodon/mastodon/pull/8127#issuecomment-411038820)), which I don't think is a particularly convincing reason given the huge time save this provides, and the push it would make to actually improve the standard and browsers' handling of it. This should be (re)introduced as an opt-in feature.
> 
> (Note that under European law, this could only ever be an opt-in feature, regardless of how nice the browser handling becomes, as it saves information to your device which is not strictly necessary for a service requested by the end user).

Yeah, during on-boarding it would be important to give users the option!

**Comment by jmaris** [comment]
Also, as we have a new influx of users, this is already causing lots of problems: the current process means people think they have to sign up to every individual instance!

![image](https://user-images.githubusercontent.com/342289/202680887-cecaa029-de9c-4fb1-8ba8-37b5b0dea7c5.png)

**Comment by arivero** [comment]
I do not think it to be the opt-in. It is the owner of the server the one who decides to activate the handler. 

The UX scare could be addressed via a mix of pathways: Keep the big COPY button, put under a small href. Some torrent sites do it; small "magnet:" button, big "download" button.

The securiry issue has not been mentioned. Of course other application could register web+mastodon, and the reaction of the browser is not wel defined. Worst case, click in the link would open the fake app and it could try to ask for credentials.

**Comment by Gottox** [comment]
@arivero I don't see how the owner of the server has any stake in that. The whole feature is implemented client side and depends mostly on remote instances. Adding a server side toggle adds in my opinion just to inconsistent UX between instances. Imho it should be completely up to the user if they want to use custom protocol handlers.

**Comment by Gottox** [comment]
The security issue is a serious issue and this needs to be discussed.  If we go with the workflow I proposed, then the attacker has a ~30x100px area they control in the iframe. The [sandbox](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe?retiredLocale=de#attr-sandbox) attribute gives the host side quite good control over the permissions of the frame and it's possible to constrain an attacker to pretty much only these few pixels.

I believe there are attack vectors, but I don't think that due to the UI it's hard to fake a login mask.

**Comment by notpushkin** [comment]
“Add Example (example.instance) as an application for web+mastodon links?” would look a bit scary for the users, but we can help them. Mastodon could first show a prompt along the lines of:

> **Make example.instance your default Mastodon server?**
> This would allow you to easily like, boost and follow users on other instances.
> (y/n)

If agreed, then we call `registerProtocolHandler` and show a hint in the background to help the user click through.

**Comment by arivero** [comment]
![image](https://user-images.githubusercontent.com/43174/202873436-b9ac409c-808d-4bb6-a2b3-56023fd00797.png)
Yeah it is ugly but not ugly to be the first motivation against it

**Comment by arivero** [comment]
Also, note that as today, the intent compatible with the old registration still exists

https://XXXX/intent?uri=web%2Bmastodon%3A%2F%2Ffollow%3Furi%3D%40arivero%40mathstodon.xyz

still works. So if you want to experiment and see how ugly the UX would be, you can register
navigator.registerProtocolHandler('web+mastodon','intent?uri=%s')
and then navigate to, say
web+mastodon://follow?uri=@arivero@mathstodon.xyz

**Comment by Gottox** [comment]
> first, the js calls web+mastodon://whoareyou? to get the hostname of the default mastodon server of the user

1. That's worse as it would leak mastodon instances to any website that request them
2. That doesn't work. You cannot simply call custom protocol handlers from Javascript.

It's pretty important, that the host site does not retrieve any user information without interaction.

**Comment by veale** [comment]
Furthermore, at the moment, Mastodon instances do not need 'cookie banners' in jurisdictions like the EU because they do not retrieve information from individuals' devices not necessary for the functioning of the site (eg tracking or extraneous functionality). If they tried to retrieve this information without asking, there would be a legal need for such a consent banner under the e-Privacy directive. Opt-in is a legal requirement for this functionality.

**Comment by arivero** [comment]
> > first, the js calls web+mastodon://whoareyou? to get the hostname of the default mastodon server of the user
> 
> 1. That's worse as it would leak mastodon instances to any website that request them

Yeah the protocol for whoareyou? should involve some signature. So the host sending the command could sign it, or the answer could be sent encripted with a public key of the sender, etc. What I am afraid is that the extra steps would surely imply extra connections.

> 2. That doesn't work. You cannot simply call custom protocol handlers from Javascript.

Damn. 

Further investigation needed here? It seems that iframe insertion works, depending on CORS. Not sure about img src or script scr. Agreed, fetch seems not to work. Not sure it at the end everything is about CORS; the custom protocol handler resolves somehow as a redirect, and the redirected target must be compliant.

> 
> It's pretty important, that the host site does not retrieve any user information without interaction.

This does not happen in this idea. All the queries are about hostnames, no usernames.

**Comment by arivero** [comment]
Alternatively, how does this OStatus thing work?

https://github.com/rugk/mastodon-simplified-federation/pull/82

has it some risk of leakages?

**Comment by veale** [comment]
> This does not happen in this idea. All the queries are about hostnames, no usernames.

The host you are registered on is definitely personal data relating to the browsing user. Not just for one-person instances, but even for larger instances could effectively can be used to fingerprint and persistently track you across websites, just as e.g. installed fonts are.

**Comment by kaleb70** [comment]
I created a discussion thread on this topic earlier, not realizing that this issue already existed. Here is a copy of my proposal. Based on the conversation here, I have added a suggestion for how a server can help protect its users from malicious protocol handlers.

### Basic Idea

If you are viewing a profile on a server where you are not logged in, the Follow button would be a link that looks like this:

`web+mstdn:profile/@kalebaxon@deacon.social?follow=1`

Unfortunately, setting up the protocol handler is browser- and platform-specific.

* Android and iOS: Mastodon client apps must be configured as the protocol handler.
* Firefox, Opera, desktop Chrome/Chromium/Edge: Use `registerProtocolHandler` and show a friendly message to the user guiding them to accept the request.
* Desktop Safari: Someone else would need to chime in on how this works.

### The Follow Button

The Follow button would execute a different script depending on whether the user is logged in.

```mermaid
graph TD
    Start([Render Follow Button]) --> LoggedIn
    LoggedIn{Logged In?}
    LoggedIn -->|Yes| DirectFollow
    LoggedIn -->|No| IndirectFollow
    DirectFollow[Script for Direct Follow]
    DirectFollow --> Done
    IndirectFollow[Script for Indirect Follow]
    IndirectFollow --> Done
    Done([Done])
```

If the user is logged in, the Follow button's behavior does not change. The change comes when the user is _not_ logged in, in which case `window.location` is set to a URL such as `web+mstdn:profile/@kalebaxon@deacon.social?follow=1`, and an alternate UI is shown for when the protocol handler is not configured.

```mermaid
graph TD
    Start([Indirect Follow]) --> GreyOut
    GreyOut[Grey Out Follow Button] --> Link
    Link[set window.location] --> Delay
    Delay[Delay to Let New Page Load] --> ShowUI
    ShowUI[Show Alternate UI] --> Done
    Done([Done])
```

### Alternate Follow UI

The alternate Follow UI would prompt the user to enter their own Mastodon server, then redirect the user there. For example, if a user is trying to follow `@kalebaxon@deacon.social` and they enter their Mastodon server as `mastodon.social`, they would be directed to `https://mastodon.social/@kalebaxon@deacon.social?follow=1&protocolhelp=1`. This would instruct the user's own server to handle the follow request and help the user authorize the protocol handler.

### Security Risk Mitigation

One security risk with this design is that a malicious website could trick the user into authorizing it as the protocol handler for `web+mstdn`. For the web interface, a server could help protect its users against this by including this process in the rendered page:

1. The page contains a sandboxed iframe to a url such as `web+mstdn:verify?token={one-time-token}`.
2. The client then checks with the server to see if the one time token was received. If not, then the user is notified and guided through the process to correct the issue.
3. The user needs the ability to disable this check.

If doing this on every page load is undesirable, it could be done after an elapsed time instead.

**Comment by Ciantic** [comment]
> We've done this before but removed because browser support / UX was inadequate.

I think revisiting the idea could be useful. Does anyone remember what was the problem it was removed? 

Underlying problem might already be fixed in Chrome/Safari.

**Comment by rcsilva83** [comment]
I think the current remote follow UX is far from easy and convenient... I think the use of procotol handlers would greatly simplify this flow and make Mastodon much more user friendly for users coming from centralized social networks, like Twitter.

I don't know the details about the reasons for this feature had been removed in the past but I read consistent arguments from other people in previous comments.

**Comment by ncfavier** [comment]
Is there a way to restore the pre-v4 behaviour of the signed-out "follow" button (the `/users/USER/remote_follow` popup that prompts you for your handle)? It already seems a lot less confusing and more practical than the new behaviour (the "sign in or copy this URL" popup).

**Comment by BANOnotIT** [comment]
How other fedirated services should handle mastodon protocol?

**Comment by rcsilva83** [comment]
The conclusion I came after these discussions is that web standards need to improve to better support federated services. Maybe representatives from Fediverse softwares should discuss together the challenges and solutions that we currently face and take the conclusions to W3C for standardization.

**Comment by tasket** [comment]
1. ActivityPub is now an Internet standard protocol
2. It draws inspiration from the decentralization of email (also a standard)
3. Hence, a protocol handler like 'mailto:' named after the protocol (not the program) is appropriate: 
     **activitypub:** or **apub:** or **fedi:** or similar idea.

BTW, I skimmed [this document](https://www.w3.org/TR/activitypub/) at w3.org but couldn't tell if it has anything useful about this use case.  One would think they would have defined a protocol handler for their newly accepted protocol.

**Comment by rcsilva83** [comment]
> 3\. Hence, a protocol handler like 'mailto:' named after the protocol (not the program) is appropriate:
>         **activitypub:** or **apub:** or **fedi:** or similar idea.

@tasket I disagree with you in this point. If I have a Mastodon account and also a Pixelfed account, both support ActivityPub but I would prefer following a Pixelfed user with my Pixelfed account.

**Comment by tasket** [comment]
This doesn't change the fact that Mastodon is a program, not a protocol.  The protocol is ActivityPub.

Habituating users and website devs to using URL protocol syntax to refer to specific programs would probably become a dead-end (and viewed as a nuisance to much of the web technical community) after a short time.

Note also that 'mailto:' doesn't try to handle multiple accounts or servers; that's a task for whichever mail app the device's 'mailto:' refers.

FWIW, that's not to say a different type of scheme is out of the question.  Handler names might focus on the type of activity, such as mblog for microblog or imgpub for image sharing.

**Comment by Gottox** [comment]
I consider the protocol name discussion out of scope. It needs support by both browser vendors and other activepub implementations.

**Comment by rcsilva83** [comment]
> Handler names might focus on the type of activity, such as mblog for microblog or imgpub for image sharing.

@tasket I think it's a great idea!


> I consider the protocol name discussion out of scope. It needs support by both browser vendors and other activepub implementations. `activepub:` as a protocol name needs to be included in the list of [permitted schemes](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler?retiredLocale=de#permitted_schemes). Which isn't going to happen any time soon.

@Gottox , it's permitted custom handlers using `web+` preffix. So this could be something like `web+microblog:` and `web+imgpub:`, using @tasket's idea.

**Comment by jmaris** [comment]
> 

In this event you can choose not to remember the option and have two handlers

**Comment by Sdaswani** [comment]
I hate to add an idea that introduces centralization into this awesomely decentralized solution, but for the sake of growth, can discovery on Mastodon be centralized? Meaning a very thin web service that makes clicking on a federated mastodon link _just work_, i.e., when you click it, if you logged into a mastodon instance, you can then view the content via your account and engage in all the behaviors appropriately? 

Currently, when I click on a mastodon link [like this one](https://mastodon.social/@dpogue), even if I'm logged into my mastodon instance, I have a terrible user flow to actually engage with that link, i.e., I have to 1) copy a link, 2) load my mastodon instance, and then 3) paste that link into a search box on my mastodon server.

What would be better is 1) when I log into my mastodon instance, the mastodon instance registers me with the centralized web service and receives back a registration token, then when 2) I click on a newly formatted mastodon link that was shared, something like https://mastodon.dns@dpogue, the centralized service will do some magic to present a page where I can interact with the content fluidly (by basically automating all the steps above).

Not sure if this was discussed above, from what I can tell, so I thought I'd suggest it.

**Comment by jmaris** [comment]
> 

I've always considered it might be good to have a "mastodon directory" where people can look up users who wish to be publicly findable, as it could help make it easier to find new users. 

However in this case I do not think it would be appropriate to rely on a centralised service for such a vital part of mastodon: if that one server goes down it would essentially make it impossible to follow people.

Instead we should focus on the existing local solutions

**Comment by Sdaswani** [comment]
> However in this case I do not think it would be appropriate to rely on a centralised service for such a vital part of mastodon: if that one server goes down it would essentially make it impossible to follow people.

True, it would be impossible to follow people when the service is down, but it's not like keeping large-scale, global services alive is as hard as it was used to be, and also consumers understand outages sometimes happen, so it wouldn't be a event that can't be recovered from.

> Instead we should focus on what can be done to improve the webhandler UX

Does the webhandler UX also 'do the right thing' on mobile? I don't think a solution that doesn't address mobile, where a ton of user traffic exists, is going to solve the problem. A centralized solution would work on mobile (since the mastodan app can register to handle deep links).

**Comment by unusualevent** [comment]
it would be great if this supported multiple instances in the handler (if there's some way for the browser to prompt you as to which instance out of two you'd like to go to)

**Comment by jmaris** [comment]
> 

Not possible as I understand it, but you could optionally use a browser extension for multiple instances

**Comment by odbol** [comment]
It would be great if this wasn't a "web+mastodon://" protocol, but instead was added to the safelist as "mastodon://" protocol. That way iOS and Android apps could also handle the links, not just PWAs.

It seems like they are open to adding new schemes: https://html.spec.whatwg.org/multipage/system-state.html#safelisted-scheme

**Comment by BANOnotIT** [comment]
> It would be great if this wasn't a "web+mastodon://" protocol, but instead was added to the safelist as "mastodon://" protocol. That way iOS and Android apps could also handle the links, not just PWAs.

iOS and Android can handle any protocol regardless "web+" prefix. So web+mastodon:// is the one that'll have the biggest adoption without safelisting.

**Comment by rcsilva83** [comment]
I think `web+activitypub` or `activitypub` would be better, as one could follow a Mastodon account from, for example, Pixelfed.

**Comment by tasket** [comment]
Since `mailto:` is not spelled `electronicmailto:` we may want to consider a shortened name for our protocol.  For instance, `apub://`.

As for using `web+` prefix, that feels like interference to me.  If a PC or phone OS wants to bundle a default app to handle the protocol, then `web+apub://` adds a web browser to the experience forcefully and unnecessarily.  People who install an apub app are most definitely going to expect that app be used henceforth, not a browser.

**Comment by BANOnotIT** [comment]
> adds a web browser to the experience forcefully and unnecessarily
@tasket web doesn't add web browser to the experience like youtu.be links don't require you to open browser and are handled by your YouTube applications.

Prefixes are not standardized and are just a part of web browsers security model. OS and applications can register any protocols.

**Comment by trwnh** [comment]
i don't think this makes sense as a generic activitypub thing. *especially* not if you use webfinger in such links.

**Comment by rcsilva83** [comment]
> i don't think this makes sense as a generic activitypub thing. 

But if a generic term isn't used, you "break" the Fediverse concept. How would a user from Mastodon could follow a Pixelfed user? Mastodon would need to handle `pixelfed://` links... But, what if, instead of Pixelfed, is another Fediverse service that Mastodon doesn't handle yet?

> _especially_ not if you use webfinger in such links.

I didn't understand... Could you make an example?

**Comment by trwnh** [comment]
i'm not sure protocol handlers are the best for this but if you were to use one then it would not be operating on activitypub. `username@domain` is not part of activitypub. it wouldn't be mastodon's role to support every protocol handler, it would be up to mastodon-compatible projects to implement `web+mastodon` or whatever you call it. pixelfed would be acting as a mastodon-compatible protocol handler to interact with the mastodon-compatible part of the web (part of which depends on activitypub, but mastodon places several addition requirements, such as having a username in a specific format, or having a username at all).

you could in theory have something like `web+activity` which attempts to fetch the activitystreams representation, if the goal is to be able to browse the object in a generic activitystreams client? the rest of the URI would have to be the exact same as after the `https://` part. so basically just replace the `https://` with `web+activity://` and this represents https + the accept header

i think another way of doing this for web browsers would be to use `proxyUrl` with an https uri, which would load the object in the client (instance) of your choice: https://github.com/mastodon/mastodon/issues/23970#issuecomment-1455693299

basically you would load the object on your local instance and then you only have to handle links from your local instance. for example, you click a button on a website and type in your `username@domain` or even just your domain. that site does a webfinger resolution and redirects to the "authorize interaction" endpoint. that might require javascript or server-side scripting though. or a browser extension.

**Comment by rcsilva83** [comment]
Are you guys aware of [this](https://github.com/activitypub-schema/proposal) and [this](https://codeberg.org/fediverse/fediverse-ideas/issues/1) ?

**Comment by josephrocca** [comment]
> Are you guys aware of [this](https://github.com/activitypub-schema/proposal) and [this](https://codeberg.org/fediverse/fediverse-ideas/issues/1) ?

Came here to say that, as mentioned in the latter link you shared, this should *definitely* work for liking/replying/etc. too.

When someone shares a mastodon post on hackernews, for example, I should be able to open that mastodon link, then click the star button on the post, and it opens the handler-chooser, and I select the instance (assuming I've registered on multiple instances), and then it opens the post within the instance I selected, and prompts me to click the star button (or does it already if that can be done in a way that's safe - not sure how protocol handlers work exactly). Or something roughly equivalent to that, UX-wise.

Copying URLs around for this sort of thing is **extremely bad UX** (I've only just started trying to use Mastodon seriously, and I can't believe people have been putting up with this UX for so long), and is a *natural forcing function towards centralisation*, since it means that if you share a post on e.g. hackernews, it'll get more interactions if you're on a server that lots of other people are on. I.e. it encourages people to join the largest server, else they'll get fewer replies/likes/etc. due to how bad the cross-server post/user interaction UX is.

**Comment by almereyda** [comment]
The double forward slashes `//` are not needed for a complete URL, and are part of the scheme. timbl stated in 2009, that this was a mistake and could have been left out of the HTTP specification without losing functionality. There are other protocol handlers which don't use them either.

<details><summary>Reference</summary>
<p>

* [The Web's Inventor Regrets One Small Thing - The New York Times](http://web.archive.org/web/20231012102028/https://archive.nytimes.com/bits.blogs.nytimes.com/2009/10/12/the-webs-inventor-regrets-one-small-thing/) (archived version)

</p>
</details>

**Comment by tasket** [comment]
Also of note:

Web sites in general are probably avoiding adding "Share with Mastodon" links to their articles because there is no way to link to a fedi function or content that references the user's preferred app (or web instance).  I've seen a few sites do this and they all link to the instance that the website owner uses, so in almost all cases the user will be taken to a 'foreign' website and definitely not their app (if they use one) where they will be prompted to log in to their home server.  This is off-putting and also sets user expectations to tolerate phishing attempts – because these foreign-site-initiated login prompts have a verify similar form to a common phishing tactic.

This may even be a reason why journalists discussing the fediverse are the most likely to say that its too complicated for some; they are likely very aware that fedi could be an asset, but see no way to make the "Share" links work smoothly.

---

I think that updating the W3C ActivityPub spec to include a short definition of an 'apub://' or similar handler prefix would get the ball rolling in the right direction as it would be a clear signal to fedi app authors to start registering that prefix with the OS during installs and updates.

A loose convention about server names could also be included, such as what the Internet used to observe: the first part of the address like 'www.' or 'mail.' if that seems to offer further clarification. Going in the other direction may be better: Defining it in the form of 'apub+<type>://' would allow links such as 'apub+mblog://' for Mastodon-like content or 'apub+photos://' for Pixelfed-like services, etc.

**Comment by thegreatGreenstar** [comment]
A format of apub+type://user@domain?postid=# might be optimal for post types. Similar to email, and it should ensure that the domain isn’t cut off by post id; and if post is gone still gets to user or at least domain id that’s right and user is deleted (if instance supports a “local” view of others instances)

**Comment by tasket** [comment]
Why would a "Share This" link from a random website contain "user@domain"?

The point of having the protocol handler is to defer the transaction to a fedi app (or website) that already knows who you are... where you are logged in.  The user should not have to be logged into that third-party website and fill in their fedi identity (presumably in a profile) in order to make these links work.

**Comment by teohhanhui** [comment]
@tasket I think they're referring to the username and domain of the post being linked to. But not everything is Mastodon, so that's just not right.
